src/execution/interpreter_loop.rs

```diff
@@ -1403,6 +1941,308 @@ pub(super) fn run<H: HookSet>(
+                                i64::MIN
+                            } else if v1.is_infinity() {
+                                i64::MAX
+                            } else {
+                                v1.as_i64()
+                            }
+                        };
+
+                        trace!("Instruction: i64.trunc_sat_f64_s [{v1}] -> [{res}]");
+                        stack.push_value(res.into());
+                    }
+                    I64_TRUNC_SAT_F64_U => {
+                        let v1: value::F64 = stack.pop_value(ValType::NumType(NumType::F64)).into();
+                        let res = {
+                            if v1.is_nan() || v1.is_negative_infinity() {
+                                0
+                            } else if v1.is_infinity() {
+                                u64::MAX as i64
+                            } else {
+                                v1.as_u64() as i64
+                            }
+                        };
+
+                        trace!("Instruction: i64.trunc_sat_f64_u [{v1}] -> [{res}]");
+                        stack.push_value(res.into());
+                    }
+                    // See https://webassembly.github.io/bulk-memory-operations/core/exec/instructions.html#xref-syntax-instructions-syntax-instr-memory-mathsf-memory-init-x
+                    // Copy a region from a data segment into memory
+                    MEMORY_INIT => {
+                        //  mappings:
+                        //      n => number of bytes to copy
+                        //      s => starting pointer in the data segment
+                        //      d => destination address to copy to
+                        let data_idx = wasm.read_var_u32().unwrap_validated() as DataIdx;
+                        let data_init_len = store.data.get(data_idx).unwrap().data.len();
+                        let mem_idx = wasm.read_u8().unwrap_validated() as usize;
+                        let mem = store.mems.get(mem_idx).unwrap_validated();
+                        let mut n: i32 = stack.pop_value(ValType::NumType(NumType::I32)).into();
+                        let mut s: i32 = stack.pop_value(ValType::NumType(NumType::I32)).into();
+                        let mut d: i32 = stack.pop_value(ValType::NumType(NumType::I32)).into();
+                        if n < 0 || s < 0 || d < 0 {
+                            return Err(RuntimeError::MemoryAccessOutOfBounds);
+                        }
+
+                        if s as usize + n as usize > data_init_len
+                            || d as usize + n as usize > mem.data.len()
+                        {
+                            return Err(RuntimeError::MemoryAccessOutOfBounds);
+                        }
+
+                        while n != 0 {
+                            let b = store.data.get(data_idx).unwrap().data[s as usize];
+                            #[allow(unused_labels)]
```

Comment:
For the places in which `i32.store8` and `i32.load8_u` are used in other instructions, I have decided to wrap them in their own block and label it accordingly so that it's easier to reason about when checking them manually for correctness.

Reply:
Algorithms have been reworked.